org.eclipse.m2e.pde.target/src/org/eclipse/m2e/pde/target/MavenPluginSourcePathLocator.java

```diff
@@ -0,0 +1,135 @@
+		String installLocation = plugin.getModel().getInstallLocation();
+		if (installLocation != null) {
+			File file = new File(installLocation);
+			if (file.isFile()) {
+				String ext = FilenameUtils.getExtension(installLocation);
+				String baseName = FilenameUtils.getBaseName(installLocation);
+				File localFile = new File(file.getParentFile(), baseName + "-sources." + ext);
+				if (localFile.isFile()) {
+					return new Path(localFile.getAbsolutePath());
+				}
+				Artifact artifact = aquireFromTargetState(plugin);
+				if (artifact == null) {
+					artifact = aquireFromJar(file);
+				}
+				if (artifact != null) {
+					try {
+						IMaven maven = MavenPlugin.getMaven();
+						String extension = artifact.getExtension();
+						if (extension == null || extension.isEmpty()) {
+							extension = "jar";
+						}
+						org.apache.maven.artifact.Artifact resolved = maven.resolve(artifact.getGroupId(),
+								artifact.getArtifactId(), artifact.getBaseVersion(), extension, "sources",
+								maven.getArtifactRepositories(), new NullProgressMonitor());
+						return new Path(resolved.getFile().getAbsolutePath());
+					} catch (CoreException e) {
+						// source not found...
+					}
+				}
+			}
+		}
+		return null;
+	}
+
+	private Artifact aquireFromJar(File file) {
+		// search in the jar for any maven info...
+		try (JarFile jarFile = new JarFile(file)) {
+			Enumeration<JarEntry> entries = jarFile.entries();
+			while (entries.hasMoreElements()) {
+				JarEntry jarEntry = entries.nextElement();
+				String name = jarEntry.getName();
+				if (name.startsWith("META-INF/maven/")) {
+					if (name.endsWith("/pom.properties")) {
+						try (InputStream stream = jarFile.getInputStream(jarEntry)) {
+							return recoverFromProperties(stream);
+						}
+					}
+					if (name.endsWith("/pom.xml")) {
+						try (InputStream stream = jarFile.getInputStream(jarEntry)) {
+							return recoverFromXML(stream);
+						}
+					}
+				}
+			}
+		} catch (IOException e) {
+			// can't read from jar...
+		}
+		return null;
+	}
+
+	private Artifact recoverFromXML(InputStream stream) {
+		try {
+			Model model = MavenPlugin.getMavenModelManager().readMavenModel(stream);
+			return new DefaultArtifact(model.getGroupId(), model.getArtifactId(), null, model.getVersion());
+		} catch (CoreException e) {
+			// can't parse model...
+		}
+		return null;
+	}
+
+	private Artifact recoverFromProperties(InputStream inputStream) throws IOException {
+		Properties properties = new Properties();
+		properties.load(inputStream);
+		return new DefaultArtifact(properties.getProperty("groupId"), properties.getProperty("artifactId"), null,
+				properties.getProperty("version"));
+	}
+
+	private Artifact aquireFromTargetState(IPluginBase plugin) {
+		ITargetPlatformService platformService = TargetPlatformService.getDefault();
+		try {
+			TargetBundle[] bundles = platformService.getWorkspaceTargetDefinition().getBundles();
+			for (TargetBundle targetBundle : bundles) {
```

Comment:
Iterating over all bundles in the TP could be a relatively slow operation if there are many.
It would be ideal if there would be some mean to obtain the TargetBundle of a external Plugin model, but I didn't find any. So maybe it would be reasonable to maintain a corresponding Mapping within M2E that is updated when the TP changes.